wallet/db_sqlite3.c

```diff
@@ -1,11 +1,141 @@
 #include <wallet/db_common.h>
 #include "gen_db_sqlite3.c"
+#include <ccan/ccan/tal/str/str.h>
+#include <lightningd/log.h>
+#include <sqlite3.h>
+#include <stdio.h>
+
 #if HAVE_SQLITE3
 
+static const char *db_sqlite3_expand(struct db_stmt *stmt)
+{
+#if HAVE_SQLITE3_EXPANDED_SQL
+	sqlite3_stmt *s = (sqlite3_stmt*)stmt->inner_stmt;
+	const char *sql;
+	char *expanded_sql;
+	expanded_sql = sqlite3_expanded_sql(s);
+	sql = tal_strdup(stmt, expanded_sql);
+	sqlite3_free(expanded_sql);
+	return sql;
+#else
+	return NULL;
+#endif
+}
+
+static const char *db_sqlite3_fmt_error(struct db_stmt *stmt)
+{
+	return tal_fmt(stmt, "%s: %s: %s", stmt->location, stmt->query->query,
+		       sqlite3_errmsg(stmt->db->conn));
+}
+
+static bool db_sqlite3_query(struct db_stmt *stmt)
+{
+	sqlite3_stmt *s;
+	sqlite3 *conn = (sqlite3*)stmt->db->conn;
+	int err;
+
+	err = sqlite3_prepare_v2(conn, stmt->query->query, -1, &s, NULL);
+
+	for (size_t i=0; i<stmt->query->placeholders; i++) {
+		struct db_binding *b = &stmt->bindings[i];
+
+		/* sqlite3 uses printf-like offsets, we don't... */
+		int pos = i+1;
+		switch (b->type) {
+		case DB_BINDING_UNINITIALIZED:
+			db_fatal("DB binding not initialized: position=%zu, "
+				 "query=\"%s\n",
+				 i, stmt->query->query);
+		case DB_BINDING_UINT64:
+			sqlite3_bind_int64(s, pos, b->v.u64);
+			break;
+		case DB_BINDING_INT:
+			sqlite3_bind_int(s, pos, b->v.i);
+			break;
+		case DB_BINDING_BLOB:
+			sqlite3_bind_blob(s, pos, b->v.blob, b->len,
+					  SQLITE_TRANSIENT);
+			break;
+		case DB_BINDING_TEXT:
+			sqlite3_bind_text(s, pos, b->v.text, b->len,
+					  SQLITE_TRANSIENT);
+			break;
+		case DB_BINDING_NULL:
+			sqlite3_bind_null(s, pos);
+			break;
+		}
+	}
+
+	if (err != SQLITE_OK) {
```

Comment:
it's a bit weird semantically to be checking if the 'err' is OK. Consider renaming to `result`?